apps/kg/train_pytorch.py

```diff
@@ -90,52 +68,78 @@ def train(args, model, train_sampler, rank=0, rel_parts=None, valid_samplers=Non
-                sample_time, forward_time, backward_time, update_time))
+            print('[{}]sample: {:.3f}, forward: {:.3f}, backward: {:.3f}, update: {:.3f}'.format(
+                rank, sample_time, forward_time, backward_time, update_time))
             sample_time = 0
             update_time = 0
             forward_time = 0
             backward_time = 0
             start = time.time()
 
-        if args.valid and step % args.eval_interval == 0 and step > 1 and valid_samplers is not None:
-            start = time.time()
-            test(args, model, valid_samplers, mode='Valid')
-            print('test:', time.time() - start)
+        if args.valid and (step + 1) % args.eval_interval == 0 and step > 1 and valid_samplers is not None:
+            valid_start = time.time()
+            if args.strict_rel_part:
+                model.writeback_relation(rank, rel_parts)
+            # forced sync for validation
+            if barrier is not None:
+                barrier.wait()
+            test(args, model, valid_samplers, rank, mode='Valid')
+            print('test:', time.time() - valid_start)
+            if barrier is not None:
+                barrier.wait()
+
+    print('train {} takes {:.3f} seconds'.format(rank, time.time() - start))
+    if args.async_update:
+        model.finish_async_update()
+    if args.strict_rel_part:
+        model.writeback_relation(rank, rel_parts)
 
-@thread_wrapped_func
 def test(args, model, test_samplers, rank=0, mode='Test', queue=None):
-    if args.num_proc > 1:
-        th.set_num_threads(4)
-
     if len(args.gpu) > 0:
         gpu_id = args.gpu[rank % len(args.gpu)] if args.mix_cpu_gpu and args.num_proc > 1 else args.gpu[0]
     else:
         gpu_id = -1
 
+    if args.strict_rel_part:
+        model.load_relation(th.device('cuda:' + str(gpu_id)))
+
     with th.no_grad():
         logs = []
         for sampler in test_samplers:
             count = 0
             for pos_g, neg_g in sampler:
-                with th.no_grad():
-                    model.forward_test(pos_g, neg_g, logs, gpu_id)
+                model.forward_test(pos_g, neg_g, logs, gpu_id)
 
         metrics = {}
         if len(logs) > 0:
             for metric in logs[0].keys():
                 metrics[metric] = sum([log[metric] for log in logs]) / len(logs)
         if queue is not None:
-            queue.put(metrics)
+            queue.put(logs)
         else:
             for k, v in metrics.items():
-                print('{} average {} at [{}/{}]: {}'.format(mode, k, args.step, args.max_step, v))
+                print('[{}]{} average {} at [{}/{}]: {}'.format(rank, mode, k, args.step, args.max_step, v))
     test_samplers[0] = test_samplers[0].reset()
     test_samplers[1] = test_samplers[1].reset()
+
+@thread_wrapped_func
+def train_mp(args, model, train_sampler, valid_samplers=None, rank=0, rel_parts=None, barrier=None):
+    if args.num_proc > 1:
+        th.set_num_threads(4)
```

Comment:
I guess we need to make it more automatic. This setup is for multi-GPU training on p3 instances. It's not optimal for multi-CPU training or other GPU instances.

Reply:
i c. Maybe we can provide a arg and give automatic value for multi-CPU and GPU.